# Pandas Playground

We will start by loading the necessary libraries we will be using.

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib as plt
import jinja2 as jnj

Now that we have imported our libraries let's lead python to the .csv files we will be using <br>
Notice that we are creating variables for each of the pathways, this makes it easy to call later.

In [2]:
orders = '/Users/markhinojosa/coffee-shop/Raw Data/ORDERS.csv'
products = '/Users/markhinojosa/coffee-shop/Raw Data/PRODUCTS.csv'
customers = '/Users/markhinojosa/coffee-shop/Raw Data/CUSTOMERS.csv'

We have our filepaths variables ready to go, let's feed through Pandas to create our Dataframes.<br>
[Learn More About Dataframes](https://realpython.com/pandas-dataframe/#:~:text=The%20pandas%20DataFrame%20is%20a,with%20in%20Excel%20or%20Calc.)

In [3]:
ordersDF = pd.read_csv(orders)
productsDF = pd.read_csv(products)
customersDF = pd.read_csv(customers)

## Data Population Validation
Our data has been loaded, but let's verify that is is populating correctly.

The .head() method will pull the top nth rows but the data set. <br>
The .head() method can be populated as .head(20) to determine the sample size of data you want to pull. <br>
If left empty the default value is 5 rows.

In [4]:
ordersDF.head()

Order ID  Order Date     Customer ID Product ID  Quantity
0  QEV-37451-860  2019-09-05  17670-51384-MA      R-M-1         2
1  QEV-37451-860  2019-09-05  17670-51384-MA    E-M-0.5         5
2  FAA-43335-268  2021-06-17  21125-22134-PX      A-L-1         1
3  KAC-83089-793  2021-07-15  23806-46781-OU      E-M-1         2
4  KAC-83089-793  2021-07-15  23806-46781-OU    R-L-2.5         2

Similar to .head() -- .tail() grabs the bottom nth rows from a dataframe.

In [5]:
productsDF.tail()

Product ID Coffee Type Roast Type  Size  Unit Price  Price per 100g  \
43    E-M-2.5         Exc          M   2.5      31.625          1.2650   
44    E-D-0.2         Exc          D   0.2       3.645          1.8225   
45    E-D-0.5         Exc          D   0.5       7.290          1.4580   
46      E-D-1         Exc          D   1.0      12.150          1.2150   
47    E-D-2.5         Exc          D   2.5      27.945          1.1178   

     Profit  
43  3.47875  
44  0.40095  
45  0.80190  
46  1.33650  
47  3.07395

The .sample() method behaves differently in that it will surface randomly selected rows from the dataframe. <br>
It also allows for variable input

In [6]:
customersDF.sample(5)

Customer ID         Customer Name                        Email  \
350  99358-65399-TC         Joshuah Awdry          jawdry9q@utexas.edu   
812  45009-09239-IV          Micky Glover           mglovermk@cnbc.com   
873  16809-16936-WF  Modesty MacConnechie  mmacconnechieo9@reuters.com   
755  24972-55878-KX      Foster Constance       fconstancekz@ifeng.com   
716  20077-67239-EC    Selestina Greedyer    sgreedyerjw@parallels.com   

            Phone Number       Address Line 1              City  \
350    +1 (318) 747-7610  7961 Blackbird Road        Shreveport   
812   +44 (898) 129-9218   95 Grasskamp Point          Burnside   
873    +1 (304) 620-6008     526 Onsgard Park        Charleston   
755    +1 (214) 388-6754  2236 Mitchell Trail            Dallas   
716  +353 (388) 882-1500  9707 Leroy Junction  Castlebellingham   

            Country Postcode Loyalty Card  
350   United States    71137           No  
812  United Kingdom     EH52          Yes  
873   United States    25362          Yes  
755   United States    75287           No  
716         Ireland      Y34           No

## Validating Data Types
We know that we have the right data but we need to ensure that the data we will be analysing is cast correcrtly. <br>
the '.dtype' method helps with this.

In [7]:
ordersDF.dtypes

Order ID       object
Order Date     object
Customer ID    object
Product ID     object
Quantity        int64
dtype: object

In [8]:
customersDF.dtypes

Customer ID       object
Customer Name     object
Email             object
Phone Number      object
Address Line 1    object
City              object
Country           object
Postcode          object
Loyalty Card      object
dtype: object

In [9]:
ordersDF.dtypes

Order ID       object
Order Date     object
Customer ID    object
Product ID     object
Quantity        int64
dtype: object

In ordersDF it looks like we have to convert 'Order Date' into a datetime type.

In [10]:
ordersDF['Order Date'] = pd.to_datetime(ordersDF['Order Date'])

^^ Notice that above we used this syntax [DF Object].[DF Column Name] <br>
We will use this syntax quite a bit in our analysis

Let's call the .dtypes method once again to verify the change.

In [11]:
ordersDF.dtypes

Order ID               object
Order Date     datetime64[ns]
Customer ID            object
Product ID             object
Quantity                int64
dtype: object

It looks like the change took.

The .info() method gives a very high level summary of your dataframe similar, to .dtypes but will give a report of non-null values and aggregates type count and memory usage.

In [12]:
ordersDF.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 5 columns):
 #   Column       Non-Null Count  Dtype         
---  ------       --------------  -----         
 0   Order ID     1000 non-null   object        
 1   Order Date   1000 non-null   datetime64[ns]
 2   Customer ID  1000 non-null   object        
 3   Product ID   1000 non-null   object        
 4   Quantity     1000 non-null   int64         
dtypes: datetime64[ns](1), int64(1), object(3)
memory usage: 39.2+ KB


The .describe() method is great for a quick calculation of a given value to the rest of the data set.

In [13]:
ordersDF.describe()

Order Date     Quantity
count                 1000  1000.000000
mean   2020-11-06 21:07:12     3.551000
min    2019-01-02 00:00:00     1.000000
25%    2019-12-17 00:00:00     2.000000
50%    2020-12-02 12:00:00     4.000000
75%    2021-10-02 06:00:00     5.000000
max    2022-08-19 00:00:00     6.000000
std                    NaN     1.681733

This quick description of the stats of a given column are useful and will be used later.

## Selecting Rows and Columns
We may not need all of the data within a dataframe so let's look at the different ways to select specific data.

Below is the syntax for selecting a single column

In [14]:
ordersDF['Order ID']

0      QEV-37451-860
1      QEV-37451-860
2      FAA-43335-268
3      KAC-83089-793
4      KAC-83089-793
           ...      
995    RLM-96511-467
996    AEZ-13242-456
997    UME-75640-698
998    GJC-66474-557
999    IRV-20769-219
Name: Order ID, Length: 1000, dtype: object

Here is example of selecting multiple columns, notice the bracket syntax.

In [15]:
ordersDF[['Order ID', 'Quantity']]

Order ID  Quantity
0    QEV-37451-860         2
1    QEV-37451-860         5
2    FAA-43335-268         1
3    KAC-83089-793         2
4    KAC-83089-793         2
..             ...       ...
995  RLM-96511-467         1
996  AEZ-13242-456         5
997  UME-75640-698         4
998  GJC-66474-557         1
999  IRV-20769-219         3

[1000 rows x 2 columns]

Below is an example of selecting specific rows. <br> 
Note that pandas will assign a row index when you load in a data frame. <br>
The : symbol acts as 'between' and in the below snippet we are calling rows between 1 and 10

In [16]:
ordersDF[1 : 10]

Order ID Order Date     Customer ID Product ID  Quantity
1  QEV-37451-860 2019-09-05  17670-51384-MA    E-M-0.5         5
2  FAA-43335-268 2021-06-17  21125-22134-PX      A-L-1         1
3  KAC-83089-793 2021-07-15  23806-46781-OU      E-M-1         2
4  KAC-83089-793 2021-07-15  23806-46781-OU    R-L-2.5         2
5  CVP-18956-553 2021-08-04  86561-91660-RB      L-D-1         3
6  IPP-31994-879 2022-01-21  65223-29612-CB    E-D-0.5         3
7  SNZ-65340-705 2022-05-20  21134-81676-FR    L-L-0.2         1
8  EZT-46571-659 2019-01-02  03396-68805-ZC    R-M-0.5         3
9  NWQ-70061-912 2019-09-05  61021-27840-ZN    R-M-0.5         1

## Filtering
Below are method for filters data with Pandas <br>
In the example below we are surfacing rows where the order quantity was greater than 5 <br> <br>
**Note I am only using the .head method to limit to return results and it not a part of the filtering syntax.**

In [17]:
ordersDF[ordersDF['Quantity'] > 5].head()

Order ID Order Date     Customer ID Product ID  Quantity
16  VAU-44387-624 2019-03-20  99643-51048-IQ    A-M-0.2         6
17  RDW-33155-159 2019-10-19  62173-15287-CU      A-L-1         6
21  NUO-20013-488 2020-12-04  03090-88267-BQ    A-D-0.2         6
31  WOQ-36015-429 2021-09-25  51427-89175-QJ    A-D-0.5         6
32  WOQ-36015-429 2021-09-25  51427-89175-QJ    L-M-0.5         6

Because we converted 'Order Date' to datetime earlier we can now filter for dates as well

In [18]:
ordersDF[ordersDF['Order Date'] > '2022-01-01'].head()

Order ID Order Date     Customer ID Product ID  Quantity
6   IPP-31994-879 2022-01-21  65223-29612-CB    E-D-0.5         3
7   SNZ-65340-705 2022-05-20  21134-81676-FR    L-L-0.2         1
12  SZW-48378-399 2022-07-02  34136-36674-OM      R-M-1         5
14  GNZ-46006-527 2022-04-05  95875-73336-RG    L-D-0.2         3
15  FYQ-78248-319 2022-06-07  25473-43727-BY    R-M-2.5         5

Below is an example of how a column can be selected and then filtered. <br>
Looking for orders after 2022-01-01

In [19]:
ordersDF['Order Date'][ordersDF['Order Date'] > '2022-01-01'].head()

6    2022-01-21
7    2022-05-20
12   2022-07-02
14   2022-04-05
15   2022-06-07
Name: Order Date, dtype: datetime64[ns]

Same as above but with multiple columns selected and then filtered. <br>
Looking for Orders quantity greater than 4

In [20]:
ordersDF[['Order ID', 'Quantity', 'Customer ID']][ordersDF['Quantity'] > 4].head()

Order ID  Quantity     Customer ID
1   QEV-37451-860         5  17670-51384-MA
11  VQR-01002-970         5  49315-21985-BB
12  SZW-48378-399         5  34136-36674-OM
15  FYQ-78248-319         5  25473-43727-BY
16  VAU-44387-624         6  99643-51048-IQ

Below is an example of how we search for a specific value. <br>
Looking for orders that had a Product ID of 'E-D-0.5'

In [21]:
ordersDF[ordersDF['Product ID']=='E-D-0.5'].head()

Order ID Order Date     Customer ID Product ID  Quantity
6    IPP-31994-879 2022-01-21  65223-29612-CB    E-D-0.5         3
118  RFH-64349-897 2019-10-22  61954-61462-RJ    E-D-0.5         3
131  VDZ-76673-968 2020-12-31  82246-82543-DW    E-D-0.5         2
162  CBT-55781-720 2021-11-15  97855-54761-IS    E-D-0.5         3
164  BYU-58154-603 2020-12-17  51971-70393-QM    E-D-0.5         4

**It's important to note though we have done filtering, none of the datarames have been explicitely altered.** <br>
To do that you would use the following syntax: <br> 
<ins>newDF = ordersDF[['Order ID', 'Quantity', 'Customer ID']][ordersDF['Quantity'] > 4].head() <ins/> <br>
Note that this would create a new dataframe with the filters you applied.

## Aggregation methods

There are some nifty quick aggreagations we can do against our data to understand it more.

The below syntax selects the Order ID columns, filters to having Quantity greater than 5, and returns the count of rows that meet the criteria <br>
The .count() methods returns the number of rows meeting any criteria; returns all if no criteria is provided.

In [22]:
ordersDF['Order ID'][ordersDF['Quantity'] > 5].count()

175

The below syntax returns a count of every row for each column.

In [23]:
ordersDF.count()

Order ID       1000
Order Date     1000
Customer ID    1000
Product ID     1000
Quantity       1000
dtype: int64

The two statments above can be used in calculations as well. Below is example of extracting the percentage of orders that had a quantiy greater than 5.

In [24]:
ordersDF['Order ID'][ordersDF['Quantity'] > 5].count() / ordersDF['Order ID'].count()

0.175

The .sum() method can be used on INT data types to return the sum values.

In [25]:
ordersDF['Quantity'].sum()

3551

## Finding Missing Values
The below methods are used to understand NaN (missing values) presented in your data.

The **.isnull()** method when used will return the table will BOOLEAN values depending on whether NaN is present.

In [26]:
customersDF.isnull()

Customer ID  Customer Name  Email  Phone Number  Address Line 1   City  \
0          False          False  False         False           False  False   
1          False          False  False         False           False  False   
2          False          False  False         False           False  False   
3          False          False  False         False           False  False   
4          False          False   True         False           False  False   
..           ...            ...    ...           ...             ...    ...   
995        False          False  False         False           False  False   
996        False          False   True         False           False  False   
997        False          False   True         False           False  False   
998        False          False  False         False           False  False   
999        False          False   True         False           False  False   

     Country  Postcode  Loyalty Card  
0      False     False         False  
1      False     False         False  
2      False     False         False  
3      False     False         False  
4      False     False         False  
..       ...       ...           ...  
995    False     False         False  
996    False     False         False  
997    False     False         False  
998    False     False         False  
999    False     False         False  

[1000 rows x 9 columns]

If you want to check if any values are null against a dataframe the .value.any() method can be called.

In [27]:
customersDF.isnull().values.any()

True

^^ We can see that in customersDF there are NaN values that exist. <br>

We can chain the .sum() method to .isnull() to count the number of missing values for each column

In [28]:
customersDF.isnull().sum()

Customer ID         0
Customer Name       0
Email             204
Phone Number      130
Address Line 1      0
City                0
Country             0
Postcode            0
Loyalty Card        0
dtype: int64

Further we can do calculations to get the NaN percentage.

In [29]:
customersDF.isnull().sum() / customersDF.count()

Customer ID       0.000000
Customer Name     0.000000
Email             0.256281
Phone Number      0.149425
Address Line 1    0.000000
City              0.000000
Country           0.000000
Postcode          0.000000
Loyalty Card      0.000000
dtype: float64

## Treating Missing Values

Pandas has a few builtin method that we call call to treat missing values are there are generally two solutions. <br>
**Solution 1** - Fill the NaN values with data (which itself has two methodologies) <br>
**Solution 2** - Drop the missing by either column or row

Below we go any further though lets create a new DF that contains missing values.

In [45]:
nanDF = ordersDF['Quantity']

## .loc allows us to specific rows and columns we want to target for a given action
nanDF.loc[20:25] = np.NaN

## np.NaN is a method that drops values from the location specified
nanDF.loc[100:105] = np.NaN 

Now let's ensure that we have NaN values in our new dataset.

In [40]:
nanDF.isnull().values.any()

True

Great, we can see that NaN values are populated but let's ensure that there is only 12 missing values.

In [41]:
nanDF.isnull().sum()

12

In [43]:
nanDF.loc[20:25]

20   NaN
21   NaN
22   NaN
23   NaN
24   NaN
25   NaN
Name: Quantity, dtype: float64

In [44]:
nanDF.loc[100:105]

100   NaN
101   NaN
102   NaN
103   NaN
104   NaN
105   NaN
Name: Quantity, dtype: float64

Ok now that we have verified there are missing values, let's treat them.

#### Dropping missing values

First let's drop the missing values on the row level.

In [47]:
## The .dropna() method will drop any rows that contains a missing value. 
droppedDF = nanDF.dropna()

**Note: This will drop the entire row, where any NaN exists** <br>
<br>
Let's compare the dataframes.

In [64]:
## len() is a python function that returns the length of object that is passed in.
len(nanDF)

1000

In [63]:
len(droppedDF)

988

^^ We can see that 12 rows have been dropped.

If you want to drop columns with missing data use the syntax: <br>
<ins> myDF.dropna(axis=1) <ins/> <br>

This cannot be performed in a single column DF.

#### Filling missing values

There are two method of filling missing values: <br>
**Method 1** - Using approximate values <br>
**Method 2** - Using previous row's data

##### Method 1

In [70]:
nanDF.describe()

count    988.000000
mean       3.545547
std        1.682784
min        1.000000
25%        2.000000
50%        3.500000
75%        5.000000
max        6.000000
Name: Quantity, dtype: float64

The mean of this dataframe is 3.545547, let's round up to 4 and fill in the missing values.

In [76]:
## .fillna() is the method used for populating NaN values.
fillDF = nanDF.fillna(4)

##Verify popoulation by checking the location of the previously missing values.
fillDF.loc[20:25]

20    4.0
21    4.0
22    4.0
23    4.0
24    4.0
25    4.0
Name: Quantity, dtype: float64

In [77]:
fillDF.describe()

count    1000.00000
mean        3.55100
std         1.67338
min         1.00000
25%         2.00000
50%         4.00000
75%         5.00000
max         6.00000
Name: Quantity, dtype: float64

^^ We can see there is only a slight deviation to the data presented on the mean, std and 50th percentile.

##### Method 2

In [82]:
## the .ffill() method uses values from the row before the NaN to populate.
fillDF2 = nanDF.ffill()

fillDF2.loc[20:25]

20    5.0
21    5.0
22    5.0
23    5.0
24    5.0
25    5.0
Name: Quantity, dtype: float64

We can see that row 19 had a value of 5 and populated to the missing values.